Guide: set up semgrep to run continuously (CI) without Semgrep app (#634)
returntocorp/semgrep-docs: docs/semgrep-ci/manual-ci-setup.md

---
slug: manual-ci-setup
append_help_link: true
title: Manually integrating Semgrep in various CI providers
description: ""
hide_title: true
---

import MoreHelp from "/src/components/MoreHelp"

> # Manually integrating Semgrep in various CI providers

Author: @efortuna Thank you for reminding me!

Run Semgrep in your CI pipeline to scan your repository for code vulnerabilities and other issues.

There are two general steps to setting up Semgrep in your CI pipeline manually.

1. Set up the CI job or action to scan with Semgrep and receive an exit code.
2. Refine the CI job's parameters.

<!-- ![Steps to integrate Semgrep in CI manually](images/image1.png "image_tooltip") -->
*Figure 1. Steps to integrate Semgrep in CI manually.*

This guide defines a job or CI job as a script executed within a certain environment and managed by a CI provider.

> By refining a job's parameters, you are able to achieve the following goals:

Author: Will leave this comment in for now as Adam may have thoughts :)

> By refining a job's parameters, you are able to achieve the following goals:

Author: Changed "refine" to configure to simplify the wording and used "Reasons for configuring additional CI job parameters"

* **Run Semgrep on a schedule.** Run full scans on mainline branches at a least intrusive time on developer teams.
* **Run Semgrep with custom rules.** Apply rules specific to your organization's business goals and coding conventions.
* **Run Semgrep when an event triggers.** Run Semgrep when a pull or merge request (PR or MR) is created. These event triggers or event hooks are dependent on your CI provider. 
* **Run Semgrep on relevant files and blocks of code.** Configure Semgrep to ignore files and folders such as test files, configuration files, and files from other vendors.

> * **Pass a Semgrep CI job even when scans report findings.** By default, manual configurations fail when any finding is detected. Configure Semgrep to pass CI jobs even when findings are reported (a **fail open** state).

Author: Agreed. I'll fix this.

## Limitations of manually configured Semgrep CI scans 

* Findings are not tiered based on severity and actions cannot be undertaken based on severity. This means that either any finding will cause the job to fail, or any finding still allows the job to pass.
* Findings are dumped to a log and are not tracked over time, so there is no record of a finding's state, such as opened, closed, or ignored.

## Setting up the CI job

### Introduction

>  `semgrep ci` is the command used to run Semgrep in a CI environment. In most cases, this is the recommended command to run in the CI job. It is a subset of the `semgrep scan` command. Features of this command include:

Author: I like "it has the following characteristics." I'll wait on suggestions from Adam as well. Thanks!

 
 * `semgrep ci` is git-aware. This means it is able to detect branches and git states. 
 * `semgrep ci` makes use of environment variables to configure its behavior. 
 * `semgrep ci` performs a diff-aware scan by default. This means it only scans changes in files when run on a pull or merge request.
 * `semgrep ci` can be run on a local `.git` repository at any time to test its behavior before running it within a CI environment.

:::note

> An alternative method of running Semgrep in your CI is to run the `semgrep scan` command, passing rules to scan, for example `semgrep scan --config auto` which performs a full scan of the repository every time the job is run.

Author: Agree! Have added it in.

:::

### GitHub Actions

To add Semgrep into your GitHub Actions pipeline, do the following steps.

> 1. Create a `semgrep.yaml` configuration file to add a Semgrep command as part of your pipeline. Refer to the [GitHub Actions code snippet](#github-actions-code-snippet).

Author: Totally valid. Additionally, you might have noticed that the terms differ across CI providers (sometimes I say "command" or "step"). This is because the CI providers use different terms, and I try to keep it in context with those terms.

> 2. Optional: You can also review the [GitHub Actions workflow syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions) to help with editing the configuration file.

Author: This is for an entirely manual set up (not sure if that was clear? Let me know.) I am trying to only include environment variables relevant to manual non-App setups.

There's this part in our giant sample config where "If you are not using Semgrep App, uncomment <ThisEnvVar>, but if you are using App, uncomment this <TokenVar>and comment out <this bunch of vars>. It's easy to comment out the wrong thing inadvertently, so instead of making the user puzzle it out, I separate out manual and App-connected CI jobs, including the config file. I hope this makes sense?

> 2. Optional: You can also review the [GitHub Actions workflow syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions) to help with editing the configuration file.

Author: I have kept the informational comments. Resolving this.

> 3. Commit the configuration file into the `./github/workflows` folder within the target repository. The Semgrep job starts automatically upon detecting the `semgrep.yaml` commit in the previous step

Author: When I tested it on GHA, the committing of the job itself starts the scan. Maybe because of push events? Hi @holdenmcg, do you know if the job starts as soon as the file is committed _for all providers?_ I may have just assumed this for GHA.

> 3. Commit the configuration file into the `./github/workflows` folder within the target repository. The Semgrep job starts automatically upon detecting the `semgrep.yaml` commit in the previous step

Reviewer: Regarding the committing starting the job; if the file is configured to run on 'push' events, then it will trigger the job when you commit the file. 
Some providers do this automatically regardless of whether the push event is present (since you're changing the CI config, so it automatically knows to re-run the job) but this isn't always the case.

> 3. Commit the configuration file into the `./github/workflows` folder within the target repository. The Semgrep job starts automatically upon detecting the `semgrep.yaml` commit in the previous step

Author: (Thanks Holden! Sorry for the late acknowledgement. I'm updating the steps.)

4. Optional: Re-run or view the job from the GitHub Actions interface.

> 5. Customize the job's behavior, such as its rules to scan and files to ignore.

Author: This is for an entirely manual set up (not sure if that was clear? Let me know.) I am trying to only include environment variables relevant to manual non-App setups.

Unless you mean that we want to make a page (without having to login to Semgrep App) dedicated to creating these configurations?

> 5. Customize the job's behavior, such as its rules to scan and files to ignore.

Reviewer 2: ah, nope. I misunderstood when I was reviewing. Carry on!

      - semgrep-scan
```

### Buildkite

To add Semgrep into your CircleCI pipeline, do the following steps.

1. Create or edit a `pipelines.yml` configuration file to add a Semgrep command as part of your pipeline. Refer to the [BuildKite code snippet](#buildkite-code-snippet). This configuration file can also be stored within Buildkite.
2. Optional: You can also review the [Buildkite steps definition](https://buildkite.com/docs/pipelines/defining-steps) to help with editing the configuration file.
3. If you are using Buildkite to store the configuration, save the file. Otherwise, commit the configuration file into the `/.buildkite` folder within the target repository.
4. The Semgrep job starts automatically upon detecting the `pipelines.yml` commit in the previous step.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note
From Buildkite's main page, click **Pipelines > ➕ button** to perform these steps within Buildkite's UI.
:::

#### Buildkite code snippet

```yaml
- label: ":semgrep: Semgrep"
  commands:
    - export SEMGREP_RULES='p/default' 
    - semgrep ci 
  
  plugins:
    - docker#v3.7.0:
        image: returntocorp/semgrep
```

> 

Author: This is not a super nit at all!! I gave the provider order a lot of thought (popularity of providers, etc.) The order of the providers here matches the order of providers in the App. And Sample CI configs is going for a major overhaul and is likely to be renamed into "Automating code scans in CI with Semgrep App". As with this doc, each provider will have their own code snippet.

### Running Semgrep in CI with other CI providers

Do one of the following methods to run Semgrep in CI with other CI providers.

#### Direct docker usage 

Reference or add the [returntocorp/semgrep](https://hub.docker.com/r/returntocorp/semgrep) Docker image directly. The method to add the Docker image varies based on the CI provider.

#### Install `semgrep`

If you cannot use the Semgrep docker image, do the following steps.

1. Run `pip install semgrep` inside the container or CI environment
2. Run `semgrep ci --config auto`.

Your customization options with other CI providers vary depending on working environment variables.

## Refining the CI job

The following sections describe methods to customize your CI job.

### Passing the CI job 

By default, Semgrep CI exits with exit code 1 if the scan returns any findings. This causes the job to fail.

Semgrep provides a **fail open** option. This enables you to suppress findings or internal Semgrep errors that block your pipeline.  Use any of the following commands:

<dl>
	<dt><code>semgrep ci</code></dt>

> 	<dd>The Semgrep CI job<strong>fails</strong> on blocking findings or on internal errors.</dd>

Reviewer 2: ```suggestion
	<dd>The Semgrep CI job <strong>fails</strong> on blocking findings or on internal errors.</dd>
```

	<dt><code>semgrep ci || [ $? != 1 ]</code></dt>

> 	<dd>The Semgrep CI job <strong>fails</strong> on blocking findings, but <strong>passes</strong> on internal errors. </dd>

Reviewer 2: ```suggestion
	<dd><strong>Fail</strong> on blocking findings, but <strong>passes</strong> on internal errors. </dd>
```

	<dt><code>semgrep ci || true</code></dt>

> 	<dd>The Semgrep CI job <strong>passes</strong> on blocking findings and on internal errors. </dd>

Reviewer 2: ```suggestion
	<dd><strong>Pass</strong> on blocking findings and on internal errors. </dd>
```

</dl>

Refer to [Semgrep exit codes](../cli-reference/#exit-codes) to understand various internal issues that cause Semgrep to fail.

### Setting a scan schedule

The following table is a summary of methods and resources to set up schedules for different CI providers.

TODO

### Customizing rules and rulesets

#### Exploring Semgrep Registry for useful rulesets

> `semgrep ci` accepts a list of rules and rulesets to run on each scan. The list is delimited by a space (` `) if the variable is exported from a command or script block, such as in the BitBucket Pipeline example. The list can also be delimited by a newline if the `SEMGREP_RULES` variable is declared through YAML syntax.

Reviewer 2: this whole paragraph is confusing to me. I don't know what you're trying to say., Maybe the examples will help, and then I can suggest alternative wording. or maybe this enough of a nudge for you to think of alternatives :-)

"such as in the BitBucket Pipeline example." -> link?

Example snippet for space-delimited list using `export SEMGREP_RULES`:

Example snippet for newline-delimited list using `SEMGREP_RULES` YAML syntax:

#### Writing your own rules

Write custom rules to enforce your team's coding standards and security practices. Rules can be forked from existing community-written rules.

See [Writing rules](https://semgrep.dev/docs/writing-rules/overview/) to learn how to write custom rules.

TODO

### Ignoring files

By default Semgrep CI skips files and directories such as tests/, node_modules/, and vendor/. It uses the same default .semgrepignore as the CLI, which can be found in the [CLI Reference](https://semgrep.dev/docs/cli-reference/#ignoring-files). This is used by Semgrep CI when no explicit .semgrepignore file is found in the root of your project.

> Use a `.semgrepignore` file to exclude files and directories from a scan. The `.semgrepignore` file follows `.gitignore` syntax. You can exclude vendored code and code for tests through this feature.

Reviewer 2: is this an additional .semgrepignore from the previous paragraph? Or should this paragraph be above the previous one?

Author: I think I should have added something like "You can make a .semgrepignore file of your own. When Semgrep detects a user-defined .semgrepignore file within the root of your repository, it stops using the default .semgrepignore file above." 

Does that help?

Reviewer 2: yes, but where does the default .semgrepignore file live?

Author: (Added it in. Resolving!)

---
slug: manual-ci-setup
append_help_link: true
title: Manually integrating Semgrep in various CI providers
description: ""
hide_title: true
---

import MoreHelp from "/src/components/MoreHelp"

# Manually integrating Semgrep in various CI providers

Run Semgrep in your CI pipeline to scan your repository for code vulnerabilities and other issues.

There are two general steps to setting up Semgrep in your CI pipeline manually.

1. Set up the CI job or action to scan with Semgrep and receive an exit code.
2. Refine the CI job's parameters.

> <!-- ![Steps to integrate Semgrep in CI manually](images/image1.png "image_tooltip") -->

Author: Not actionable for reviewers. (Under construction, commented out for now.)

:::

### GitHub Actions

To add Semgrep into your GitHub Actions pipeline, do the following steps.

1. Create a `semgrep.yaml` configuration file to add a Semgrep command
   as part of your pipeline. Refer to the [GitHub Actions code
   snippet](#github-actions-code-snippet).
2. Optional: You can also review the [GitHub Actions workflow syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions) to help with editing the configuration file.
3. Commit the configuration file into the `./github/workflows` folder within the target repository. The Semgrep job starts automatically upon detecting the `semgrep.yaml` commit in the previous step
4. Optional: Re-run or view the job from the GitHub Actions interface.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note
If you are self-hosting your repository, you must [use a self-hosted runner](https://docs.github.com/en/actions/using-jobs/choosing-the-runner-for-a-job#choosing-self-hosted-runners).
:::

#### GitHub Actions code snippet

> ```yaml

Author: I don't know if the excessive(?) whitespace here helps or hinders. Maybe @efortuna and @holdenmcg have a suggestion for easier readability?

Author: I was thinking we want a config file that matches what the user wants to do. Not all users want to run a CI job with App. Some environment variables are only for App-connected scanning. Basically, I went through all the "Large Configs" and separated out only what applies to the set up method the user wants.

1. Edit or create your `Jenkinsfile` configuration file to add a Semgrep `step` as part of your Pipeline. Refer to the [Jenkins CI code snippet](#jenkins-ci-code-snippet). You can edit your `Jenkinsfile` from Jenkins's interface.
2. Optional: You can also review [`Jenkinsfile` usage](https://www.jenkins.io/doc/book/pipeline/jenkinsfile/) to help with editing the configuration file.
3. If you are using Jenkins Classic UI, [save your `Jenkinsfile`](https://www.jenkins.io/doc/book/pipeline/getting-started/#through-the-classic-ui) within the Jenkins interface. Otherwise, commit the configuration file into the root folder within the target repository. The Semgrep job starts automatically upon detecting the `Jenkinsfile` commit in the previous step. 
4. Optional: Re-run the stage from [Jenkins's interface](https://www.jenkins.io/doc/book/pipeline/running-pipelines/#restarting-from-the-classic-ui).
6. Customize the job's behavior, such as its rules to scan and files to ignore.

#### Jenkins CI code snippet

This code snippet uses Jenkins declarative syntax.

> ```javascript

Author: Jenkinsfile is resembles javascript (not yaml) so I used that for syntax highlighting. Hope this is okay. Opinions welcome.

Author: (I hope my previous responses answer this question as to why the "commented out" vars aren't here.)

   a Semgrep `step` as part of your pipeline. Refer to the [BitBucket Pipelines code snippet](#bitbucket-pipelines-code-snippet).
2. Optional: You can also review the [Bitbucket guidelines guide](https://support.atlassian.com/bitbucket-cloud/docs/configure-bitbucket-pipelinesyml/) to help with editing the configuration file.
3. Commit the configuration file into the root folder within the target repository. The Semgrep job starts automatically upon detecting the `bitbucket-pipelines.yml` commit in the previous step.
4. Optional: Re-run or view the Pipeline by clicking from **your repository > Pipelines**.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note
These steps can also be performed through BitBucket's UI wizard. This UI wizard can be accessed through **BitBucket > your repository > Pipelines > Create your first pipeline**.
:::

#### BitBucket Pipelines code snippet

```yaml
image: atlassian/default-image:latest

pipelines:
  default:
    - parallel:
      - step:
          name: 'Run Semgrep scan with current branch'
          deployment: dev
          image: returntocorp/semgrep
          script:

>             - export SEMGREP_RULES="p/default" # TODO get confirmation from Holden

Author: @holdenmcg Is this the right way to set SEMGREP_RULES in bitbucket?

Reviewer: Yes, this is exactly it.

            - semgrep ci # TODO confirm that this will work
```

### CircleCI

To add Semgrep into your CircleCI pipeline, do the following steps.

1. Create or edit a `config.yml` configuration file to add a Semgrep command as
   part of your CircleCI workflow. Refer to the [CircleCI code snippet](#circleci-code-snippet).
2. Optional: You can also review the [CirclCI YAML introduction](https://circleci.com/docs/introduction-to-yaml-configurations) to help with editing the configuration file.
3. Commit the configuration file into the `/.circleci` folder within the target
   repository.
4. The Semgrep job starts automatically upon detecting the `config.yml` commit in the previous step.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

#### CircleCI code snippet

```yaml
version: 2.1
jobs:
  semgrep-scan:
    parameters:
      repo_path:
        type: string
		    # Change This to your repository path.
			  # Sara: Isn't there a way to fetch these values from Circle CI so it's not hardcoded?
			  # TODO Check with Holden.
        default: myorg/semgrep-test-repo

>       default_branch:

Author: @holdenmcg Our current docs have this `myorg/semgrep-test-repo` - shouldn't the CI provider be feeding this such as << parameters.default_soundthing >>?

Author: Update. I checked all the parameters in the the the [CircleCI docs](https://circleci.com/docs/variables#pipeline-values) and it does not look like there's an easy way to do this (seems like I can't sed to extract the path/reponame either). @holdenmcg guess this means it's a dead end and we have to make the user do it manually?

Reviewer: Ah, I have an update for this; there is a line we can use so they don't have to set this manually. I will update the docs with the new config; we don't need them to set any kind of repo_path variable.

In this case however, you can just remove lines 283 - 286 (the entire repo_path parameter), since we're using p/default and the repo path variable is for connecting to the app. 

So I will separately update the sample CI config/app setup page with the correct info, but for now you can remove that bit.

Author: Done. Thank you Holden.

    steps:
      - checkout
      - run:
          name: "Semgrep scan"
          command: semgrep ci
workflows:
  main:
    jobs:
      - semgrep-scan
```

### Buildkite

To add Semgrep into your CircleCI pipeline, do the following steps.

1. Create or edit a `pipelines.yml` configuration file to add a Semgrep command as part of your pipeline. Refer to the [BuildKite code snippet](#buildkite-code-snippet). This configuration file can also be stored within Buildkite.
2. Optional: You can also review the [Buildkite steps definition](https://buildkite.com/docs/pipelines/defining-steps) to help with editing the configuration file.
3. If you are using Buildkite to store the configuration, save the file. Otherwise, commit the configuration file into the `/.buildkite` folder within the target repository.
4. The Semgrep job starts automatically upon detecting the `pipelines.yml` commit in the previous step.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note
From Buildkite's main page, click **Pipelines > ➕ button** to perform these steps within Buildkite's UI.
:::

#### Buildkite code snippet

```yaml
- label: ":semgrep: Semgrep"
  commands:

>     - export SEMGREP_RULES='p/default' # to confirm with Holden that this will work

Author: @holdenmcg One of the things we're trying to do is use p/default to expose the SEMGREP_RULES value, so that it's easy for folks to either migrate to semgrep app (no need to edit semgrep ci --config auto) or add more rulesets. Will the above command work with buildkite?

Reviewer: Tried this just now, works like a charm! 👍

---
slug: manual-ci-setup
append_help_link: true
title: Running Semgrep in CI without Semgrep App 
description: ""
hide_title: true
---

import MoreHelp from "/src/components/MoreHelp"

# Running Semgrep in continuous integration (CI) without Semgrep App

> Run Semgrep in your Continuous Integration (CI) pipeline to scan your repository for code vulnerabilities and other issues. This guide explains how to set up Semgrep in your pipeline without the use of Semgrep App, also known as a **stand-alone** setup.

Reviewer 3: Maybe we can also link to the section -  `## Migrating to Semgrep App from a manual CI setup`

> Run Semgrep in your Continuous Integration (CI) pipeline to scan your repository for code vulnerabilities and other issues. This guide explains how to set up Semgrep in your pipeline without the use of Semgrep App, also known as a **stand-alone** setup.

Author: This isn't actionable right now because the "Running Semgrep CI with Semgrep App" isn't merged yet (it comes after this one.) I'll do this in another PR. I prefer not to link the migrating section, it's already in the sidebar and we're loading too many links upfront.

![Steps to run Semgrep in CI without Semgrep App](/img/semgrep-ci-overview-noapp.png "Steps to integrate Semgrep in CI without Semgrep App")
*Figure 1. Steps to run Semgrep in CI manually.*

This guide defines a job or CI job as a script executed within a certain environment and managed by a CI provider.

By refining a job's parameters, you are able to achieve the following goals:

* **Run Semgrep on a schedule.** Run full scans on mainline branches at the least intrusive time on developer teams.
* **Run Semgrep with custom rules.** Apply rules specific to your organization's business goals and coding conventions.
* **Run Semgrep when an event triggers.** Run Semgrep when a pull or merge request (PR or MR) is created. These event triggers or event hooks are dependent on your CI provider. 
* **Run Semgrep on relevant files and blocks of code.** Configure Semgrep to ignore files and folders such as test files, configuration files, and files from other vendors.
* **Configure a Semgrep CI job to pass even when any finding is detected.** By default, manual configurations pass when any finding is detected. You can also configure Semgrep to fail CI jobs when findings are reported.

> * **Output, export, or save findings to a file.** Semgrep can save to a number of file formats, including SARIF and JSON. 

Reviewer 3: @s-santillan "By default, manual configurations pass when any finding is detected." - I was not sure here if that is true. I thought that all findings are blocking in these configurations but maybe this is unrelated.

## Limitations of manually configured Semgrep CI scans 

> * Findings are not tiered based on a user-defined criteria, such as the severity of a finding, and actions cannot be undertaken based on this criteria. Alternatively, Semgrep App's [Rule Board](/semgrep-app/rule-board) can tier rules and undertake actions, such as blocking PRs or MRs on a finding generated by a reliable rule.

Author: "Tiers of findings are not based..." is incorrect. The whole point is that findings are not tiered because they mechanism doesn't exist. That means that yes, any finding is blocking. There is no tier in non-app-connected scans.

Action I can undertake is to I'll make the sentences in active.

> * Findings are not tiered based on a user-defined criteria, such as the severity of a finding, and actions cannot be undertaken based on this criteria. Alternatively, Semgrep App's [Rule Board](/semgrep-app/rule-board) can tier rules and undertake actions, such as blocking PRs or MRs on a finding generated by a reliable rule.

Reviewer 3: Thank you for clarifying this. The active voice certainly helps. Maybe (just an idea) again specifying that this is not an issue with Semgrep App, but only in stand-alone setup makes sense, but this is optional and I will leave this up to you. @s-santillan

> * Findings are not tiered based on a user-defined criteria, such as the severity of a finding, and actions cannot be undertaken based on this criteria. Alternatively, Semgrep App's [Rule Board](/semgrep-app/rule-board) can tier rules and undertake actions, such as blocking PRs or MRs on a finding generated by a reliable rule.

Reviewer 3: But we specify this above the list, so feel free to ignore me here. @s-santillan

* Findings are dumped to a log, so there is no record of a finding's **triage state**, such as opened, closed, or ignored.

:::note

> `semgrep ci` prevents the duplication of findings by scanning only the changes in files in pull or merge requests. This is different from Semgrep App's capability to track a finding's **triage state**, as each finding has a record.

Author: Triage state has to do with a finding, nothing else.

1. Create a `semgrep.yaml` configuration file to add a Semgrep command as part of your Github Actions pipeline. Refer to the [GitHub Actions code snippet](#github-actions-code-snippet).
2. Optional: You can also review the [GitHub Actions workflow syntax](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions) to help with editing the configuration file.
3. Commit the configuration file into the `./github/workflows` folder within the target repository. The Semgrep job starts automatically upon detecting the `semgrep.yaml` commit in the previous step
4. Optional: Re-run or view the job from the GitHub Actions interface.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note
If you are self-hosting your repository, you must [use a self-hosted runner](https://docs.github.com/en/actions/using-jobs/choosing-the-runner-for-a-job#choosing-self-hosted-runners).
:::

#### GitHub Actions code snippet

```yaml
# Name of this GitHub Actions workflow.
name: Semgrep

on:
  # To scan changed files in PRs (diff-aware scanning):
  pull_request: {}
  # To scan mainline branches and report all findings:
  push:

>     branches: ["master", "main"]

Author: We can and that's the recommended way of running the job.

  # To schedule the CI job (this method uses cron syntax):
  schedule:
    - cron: '30 0 1,15 * *' # scheduled for 00:30 UTC on both the 1st and 15th of the month

jobs:
  semgrep:
    # User-definable name of this GitHub Actions job:
    name: Scan
    # If you are self-hosting, change the following `runs-on` value: 

>     runs-on: ubuntu-latest

Author: This is beyond the scope of this doc. The person using the config file will have the value they need for the runner.

      - run: semgrep ci
        env:
           # Define rules to scan with through the SEMGREP_RULES environment variable. 
           SEMGREP_RULES: p/default # more at semgrep.dev/explore
           # Uncomment SEMGREP_TIMEOUT to set this job's timeout (in seconds):
           # (Default is 1800 seconds. Set to 0 to disable.)
           # SEMGREP_TIMEOUT: 300
```

### GitLab CI/CD

To add Semgrep into your GitLab CI/CD pipeline:

1. Create or edit a `.gitlab-ci.yml` configuration file to add a Semgrep
   command as part of your GitLab CI/CD pipeline. Refer to the [GitLab CI/CD code
   snippet](#gitlab-cicd-code-snippet).
2. Optional: You can also review the [GitLab configuration guide](https://docs.gitlab.com/ee/ci/yaml/gitlab_ci_yaml.html) to help with editing the configuration file. 
3. Commit the configuration file into the root folder within the target repository. The Semgrep job starts automatically upon detecting the `.gitlab-ci.yml` commit in the previous step.
4. Optional: Re-run or view the job from your GitLab project's **CI/CD > Pipelines page**.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note

> If you are already running [GitLab SAST](https://docs.gitlab.com/ee/user/application_security/sast/) by including the template `Security/SAST.gitlab-ci.yml` in your CI/CD configuration, you can still include and customize Semgrep's CI integration. GitLab SAST, including its `semgrep-sast` analyzer, will continue to run normally. Refer to the [GitLab CI/CD with GitLab SAST snippet](#gitlab-cicd-with-gitlab-sast-snippet).

Author: "by including" is relevant.

4. The Semgrep job starts automatically upon detecting the `pipelines.yml` commit in the previous step.
5. Customize the job's behavior, such as its rules to scan and files to ignore.

:::note
These steps can be performed from within Buildkite's interface. From Buildkite's main page, click **Pipelines > ➕ button** to perform these steps within Buildkite's UI.
:::

#### Buildkite code snippet

```yaml
- label: ":semgrep: Semgrep"
  commands:
    # Define rules to scan with through the SEMGREP_RULES environment variable. 
    - export SEMGREP_RULES="p/default"
    # To scan changed files in PRs or MRs (diff-aware scanning):
    - export SEMGREP_BASELINE_REF=${BUILDKITE_BRANCH}
    # Uncomment SEMGREP_TIMEOUT to set this job's timeout (in seconds):
    # (Default is 1800 seconds. Set to 0 to disable.)
    # - export SEMGREP_TIMEOUT="300"
    - semgrep ci 
  
  plugins:
    - docker#v3.7.0:
      image: returntocorp/semgrep
```

### Running Semgrep through other CI providers

Use either of the following methods to run Semgrep through other CI providers.

#### Direct docker usage 

> Reference or add the [returntocorp/semgrep](https://hub.docker.com/r/returntocorp/semgrep) Docker image directly. The method to add the Docker image varies based on the CI provider.

Author: I linked to one of our existing snippets.

> #### Install `semgrep`

Author: We are not emphasizing Semgrep as a docker image in this one.

If you cannot use the Semgrep docker image, install `semgrep` as a step or command within your CI job:

1. Add `pip install semgrep` into the configuration file as a step or command, depending on your CI provider's syntax.
2. Run any valid `semgrep ci` command, such as `semgrep ci --config auto`.

See the following example using BitBucket Pipelines:

```yaml
image: atlassian/default-image:latest

pipelines:
  default:
    - parallel:
      - step:
          name: 'Run Semgrep scan with current branch'
          deployment: dev
          script:
          - pip install semgrep
            - export SEMGREP_RULES="p/default" 
            - semgrep ci
```

Your customization options with other CI providers vary depending on working environment variables.

## Refining the CI job

> The following sections describe methods to customize your CI job. Many of these features require the correct setup of environment variables.

Reviewer 3: Maybe turn the sentence toward the user. @s-santillan `You need to provide the correct setup of environment variables for many features mentioned below.`

> The following sections describe methods to customize your CI job. Many of these features require the correct setup of environment variables.

Author: Just took it out entirely since the sections already tell the user what to do.

> ### Passing or failing the CI job

Author: Already solved in another PR.

By default, a Semgrep CI job exits with exit code 1 if the scan returns any findings. This causes the job to fail.

Semgrep provides **fail open** options. These options enable you to suppress findings that block your pipeline:

<dl>
	<dt><code>semgrep ci</code></dt>
	<dd><strong>Fail</strong> on blocking findings, but <strong>passes</strong> on internal errors. This is the default behavior.</dd>
	<dt><code>semgrep ci --no suppress errors</code></dt>
	<dd>The Semgrep CI job <strong>fails</strong> on blocking findings and on internal errors.</dd>
	<dt><code>semgrep ci || true</code></dt>
	<dd><strong>Pass</strong> on blocking findings and on internal errors.</dd>
</dl>

> Refer to [Semgrep exit codes](../cli-reference/#exit-codes) to understand various internal issues that cause Semgrep to fail.

Author: "Configuration reference" is going to be removed after these docs are merged. "configuation reference" doc is a collection of features that the new docs replace by rewriting "configuration reference"'s functions into a coherent grouping. The correct thing to do is to add the section from "configuration reference" into these docs. Then delete the old doc.

### Diff-aware scanning

> To enable diff-aware scanning, Semgrep provides a `SEMGREP_BASELINE_REF` environment variable that users can set with the **commit hash** or **branch name** as the baseline. For example, given a repository with 10 commits, setting `SEMGREP_BASELINE_REF` to the 8th commit will return results introduced by changes in the 9th and 10th commit.

Author: Skipping stuff that I'm just going to copy paste from the configuration reference.

:::

### Setting a scan schedule

The following table is a summary of methods and resources to set up schedules for different CI providers.

| CI provider | Where to set schedule | Resource |
| ----------- | --------------------  | -------  |
| GitHub Actions | Within `semgrep.yml` file | [Sample code snippet](#github-actions-code-snippet) |
| GitLab CI/CD | Within GitLab CI/CD interface |[Official documentation](https://docs.gitlab.com/ee/ci/pipelines/schedules.html) |
| Jenkins | Within Jenkins interface |  [Official documentation](https://www.jenkins.io/doc/book/pipeline/running-pipelines/#scheduling-jobs-in-jenkins) |
| BitBucket Pipelines | Within BitBucket Pipelines interface | [Official documentation](https://support.atlassian.com/bitbucket-cloud/docs/pipeline-triggers/) |
| CircleCI | Within CircleCI interface | [Official documentation](https://circleci.com/docs/scheduled-pipelines#get-started-with-scheduled-pipelines-in-circleci) |
| Buildkite | Within Buildkite interface | [Official documentation](https://buildkite.com/docs/pipelines/scheduled-builds) |

### Customizing rules and rulesets

> #### Adding rules to scan with `semgrep ci`

Reviewer 3: Note: Great section!

`semgrep ci` accepts a list of rules and rulesets to run on each scan. The rules and rulesets can come from the [Semgrep Registry](https://semgrep.dev/r/), or your own rules. The sources for rules to scan with are:

* A `.semgrep` folder located at the root of your repository.
* The value of the `SEMGREP_RULES` environment variable.

The `SEMGREP_RULES` environment variable accepts a list of local and remote rules and rulesets to run. The `SEMGREP_RULES` list is delimited by a space (` `) if the variable is exported from a shell command or script block. For example, see the following BitBucket Pipeline snippet:

```yaml
# ...
  script:
    - export SEMGREP_RULES="p/nginx p/ci no-exec.yml" 
    - semgrep ci
# ...
```

The line defining `SEMGREP_RULES` defines three difference sources, delimited by a space:

```
- export SEMGREP_RULES="p/nginx p/ci no-exec.yml" 
```

> The example references two rulesets from Semgrep Registry (`p/nginx` and `p/ci`)and a rule available in the repository (`no-exec.yml`).

Reviewer 3: ```suggestion
The example references two rulesets from Semgrep Registry (`p/nginx` and `p/ci`) and a rule available in the repository (`no-exec.yml`).
```

If the `SEMGREP_RULES` environment variable is defined from a YAML block, the list of rules and rulesets to run is delimited by a newline. See the following example of a GitLab CI/CD snippet:
```YAML
# ...
variables:
  SEMGREP_RULES: >-
    p/nginx
    p/ci
    no-exec.yml
# ...
```

#### Writing your own rules

Write custom rules to enforce your team's coding standards and security practices. Rules can be forked from existing community-written rules.

See [Writing rules](https://semgrep.dev/docs/writing-rules/overview/) to learn how to write custom rules.

### Ignoring files

> By default `semgrep ci` skips files and directories such as `tests/`, `node_modules/`, and `vendor/`. It uses the default `.semgrepignore` file which can be found in the [Semgrep GitHub repository](https://github.com/returntocorp/semgrep/blob/develop/.semgrepignore). This default is used when no explicit `.semgrepignore` file is found in the root of your repository.

Reviewer 3: ```suggestion
By default `semgrep ci` skips files and directories such as `tests/`, `node_modules/`, and `vendor/`. It uses the default `.semgrepignore` file which you can find in the [Semgrep GitHub repository](https://github.com/returntocorp/semgrep/blob/develop/.semgrepignore). This default is used when no explicit `.semgrepignore` file is found in the root of your repository.
```

> You can copy and commit the default `.semgrepignore` file to the **root of your repository** and extend it with your own entries or write one from scratch. If Semgrep detects a `.semgrepignore` file within your repository, it won't append entries from the default `.semgrepignore` file.

Reviewer 3: ```suggestion
You can copy and commit the default `.semgrepignore` file to the **root of your repository** and extend it with your own entries or write one from scratch. If Semgrep detects a `.semgrepignore` file within your repository, it does not append entries from the default `.semgrepignore` file.
```

For a complete example, see the [.semgrepignore file on Semgrep’s source code](https://github.com/returntocorp/semgrep/blob/develop/.semgrepignore).

:::caution
`.semgrepignore` is only used by Semgrep CI and the Semgrep command line tool. Integrations such as [GitLab's Semgrep SAST Analyzer](https://gitlab.com/gitlab-org/security-products/analyzers/semgrep) do not use it.
:::

For information on ignoring individual findings in code, see the [Ignoring findings page](https://semgrep.dev/docs/ignoring-findings/).

### Saving or exporting findings to a file

> To save or export findings, you can pass file format options and send the formatted findings to a file.

Reviewer 3: Optional and non-blocking. We do not have to express the "you can" always (but we can of course, just to make it read a bit better here).

```suggestion
To save or export findings, pass file format options and send the formatted findings to a file.
```

For example, to save to a JSON file:

`semgrep ci --json --output findings.json`

You can also use output redirection (`>`):

`semgrep ci --sarif > findings.sarif`

> Refer to the [CLI reference](/cli-reference) for output formats.

Reviewer 3: Note: Thanks for documenting this so nicely and shortly!

> ## Migrating to Semgrep App from a manual CI setup

Reviewer 3: I see, the section I recommended above is here - maybe we can link here from the Limitations section above? Just an idea.

Author: I prefer that we link to the actual Running Semgrep CI in Semgrep App doc, once it's available.

Migrate to Semgrep App to:

* **View and manage findings in a centralized location**. False positives can be ignored.

> * **Configure rules and actions to undertake when a finding is generated by the rule.** The following actions can be undertaken:

Reviewer 3: ```suggestion
* **Configure rules and actions to undertake when a finding is generated by the rule**. You can undertake the following actions:
```

:::note
These steps can be performed from within Buildkite's interface. From Buildkite's main page, click **Pipelines > ➕ button** to perform these steps within Buildkite's UI.
:::

#### Buildkite code snippet

```yaml
- label: ":semgrep: Semgrep"
  commands:
    # Define rules to scan with by setting the SEMGREP_RULES environment variable. 
    - export SEMGREP_RULES="p/default"
    # To scan changed files in PRs or MRs (diff-aware scanning):
    - export SEMGREP_BASELINE_REF=${BUILDKITE_BRANCH}
    # Uncomment SEMGREP_TIMEOUT to set this job's timeout (in seconds):
    # Default timeout is 1800 seconds (30 minutes).
    # Set to 0 to disable the timeout.
    # - export SEMGREP_TIMEOUT="300"
    - semgrep ci 
  
  plugins:
    - docker#v3.7.0:
      image: returntocorp/semgrep
```

### Running Semgrep through other CI providers

Use either of the following methods to run Semgrep through other CI providers.

#### Direct docker usage 

Reference or add the [returntocorp/semgrep](https://hub.docker.com/r/returntocorp/semgrep) Docker image directly. The method to add the Docker image varies based on the CI provider.

> #### Install `semgrep`

Reviewer 3: @s-santillan maybe we can change this header to be a bit more specific, so folks will not find it when searching for general "how to install Semgrep".

Author: Done in latest push.

### Ignoring files

By default `semgrep ci` skips files and directories such as `tests/`, `node_modules/`, and `vendor/`. It uses the same default `.semgrepignore` as the `semgrep` command. This default `semgrepignore`  can be found in the [CLI Reference](https://semgrep.dev/docs/cli-reference/#ignoring-files). This is used by `semgrep ci` when no explicit `.semgrepignore` file is found in the root of your repository.

> You can copy and commit the default `.semgrepignore` and extend it with your own entries or write one from scratch. `.semgrepignore` follows `.gitignore` syntax. If Semgrep detects a `.semgrepignore` file within your repository, it won't append entries from the default `.semgrepignore` file.

Reviewer 2: "If Semgrep detects a `.semgrepignore` file within your repository"
This makes it sound a little like you can place your own .semgrepignore literally anywhere in your repo? Is that true? These two paragraphs are confusing about placement of the default an your own custom .semgrepignore.

Author: I have added its location (at the root of the repository). Thank you @efortuna!
